slack_wrapper.gs

```diff
@@ -0,0 +1,184 @@
+      var payload = par.payload;
+      if (payload){ // if payload exists, this is a doPost event from a slack interactive component
+        this.parseEventInteractiveMessage(JSON.parse(payload));
+      } else{ // else, this is a doPost event from a slack slash command
+        this.parseEventSlashCommand(par);
+      }
+    }    
+  }
+  
+  parseEventInteractiveMessage(par){
+    this.token = par.token;
+    this.teamid = par.team.id;
+    
+    this.type = par.type; 
+    this.subtype = par.view.callback_id;
+    
+    this.channelid = view.private_metadata.channelid;
+    this.userid = par.user.id;
+    this.response_url = view.private_metadata.response_url;
+    
+    this.args.uniqueid = view.private_metadata.uniqueid;
+    this.args.more = view.state.values;    
+  }
+  
+  
+  parseEventSlashCommand(par){
+    this.token = par.token;
+    this.teamid = par.team_id;
+    
+    this.type = 'command';
+    this.subtype = par.command;
+    
+    this.channelid = par.channel_id;
+    this.userid = par.user_id;
+    this.username = par.user_name;
+    this.response_url = par.response_url;
+    this.trigger_id = par.trigger_id
+    
+    this.argsString=par.text;
+    this.parseEventSlashCommandArgs(par.text); // populate this.args
+  }
+  
+  parseEventSlashCommandArgs(txt){
+    // parse txt string and store parsed values in this.args
+    if(txt) {
+      var args = txt.split(' ');
+      this.args.uniqueid = args[0]; // if uniqueid is specified, it is always the first argument
+      this.args.mentionName = args[1]; // if mentionName is specified, it is always the second argument
+    }
+  }
+  
+  
+  checkAuthenticity(){
+    // fetch validation variables
+    var globvar = globalVariables();
+    var teamid_true =  globvar['TEAM_ID'];
+    var token_true = PropertiesService.getScriptProperties().getProperty('VERIFICATION_TOKEN'); // expected verification token that accompanies slack API request     
+  
+    // initialise return message
+    var output = {code:false, msg:''};
+    
+    // check token
+    if(!token_true){ // check that token_true has been set in script properties
+      output.msg = 'error: VERIFICATION_TOKEN is not set in script properties. The command will not run. Please contact the web app developer.';
+      return(output);
+    }  
+    if (this.token !== token_true) {
+      output.msg = 'error: Invalid token ' + this.token + ' . The command will not run. Please contact the web app developer.';
+      return(output);
+    }
+    
+    // check request originates from our slack workspace
+    if (this.teamid != teamid_true){
+      output.msg = 'error: You are sending your command from an unauthorised slack workspace.';
+      return(output);
+    }
+    
+    output.code=true;
+    return (output);
+  }
+  
+  
+  checkArgSyntax(){
+    //**** todo ****//
+  }
+  
+  handleEvent (){
+    //  handle slack doPOST events
+  
+    // extract relevant data from message body
+    var workspace = this.teamid;
```

Comment:
I thought it would be better practice to create intermediate variables for elements of the object that are called multiple times in the subfunctions? In case we restructure/rename class objects, we would just need to update in one place in that function. Or maybe there's a better reason not to?

That being said I want to simply the command handling soon, so I may revise this anyway.

Reply:
This is usually only the case if you are going to do some intermediate transformation on an element. In this case you are passing all of them through directly, so I think just using `this.thing` is fine. 

You're right that if we changed the interface, we would have to change thing, but I'd rather add that in if and when we need it.

Reply:
okay, I'm refactoring this currently:
- Created a key-value globvar object to match slack command to intended function name
- A one liner to call function + args rather than all the if statements.
- All subcommands (volunteer, cancel, etc.) take a single object as argument which contains all the relevant args for that function (args.uniqueid, args.channelid, etc.)